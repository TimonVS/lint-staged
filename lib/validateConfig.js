/* eslint no-prototype-builtins: 0 */

'use strict'

const debug = require('debug')('lint-staged:cfg')

const { configurationError, incorrectBraces } = require('./messages')

const TEST_DEPRECATED_KEYS = new Map([
  ['concurrent', (key) => typeof key === 'boolean'],
  ['chunkSize', (key) => typeof key === 'number'],
  ['globOptions', (key) => typeof key === 'object'],
  ['linters', (key) => typeof key === 'object'],
  ['ignore', (key) => Array.isArray(key)],
  ['subTaskConcurrency', (key) => typeof key === 'number'],
  ['renderer', (key) => typeof key === 'string'],
  ['relative', (key) => typeof key === 'boolean'],
])

/**
 * Braces with a single value like `*.{js}` are invalid
 * and thus ignored by micromatch. This regex matches all occurrences of
 * two curly braces without a `,` or `..` between them, to make sure
 * users can still accidentally use them without
 * some linters never matching anything.
 *
 * For example `.{js,ts}` or `file_{1..10}` are valid but `*.{js}` is not.
 *
 * @see https://www.gnu.org/software/bash/manual/html_node/Brace-Expansion.html
 */
const BRACES_REGEXP = /({)(?:(?!,|\.\.).)*?(})/g

/**
 * Remove braces from incorrect glob patterns.
 * For example `*.{js}` is incorrect because it doesn't contain a `,` or `..`,
 * and will be reformatted as `*.js`.
 *
 * @param {string} pattern the glob pattern
 * @returns {string}
 */
const withoutIncorrectBraces = (pattern) => {
  let output = `${pattern}`
  while (BRACES_REGEXP.exec(pattern)) {
    output = output.replace(/{/, '')
    output = output.replace(/}/, '')
  }
  return output
}

/**
 * Runs config validation. Throws error if the config is not valid.
 * @param config {Object}
 * @returns config {Object}
 */
const validateConfig = (config, logger) => {
  debug('Validating config')

  if (!config || typeof config !== 'object') {
    throw new Error('Configuration should be an object!')
  }

  const errors = []

  if (Object.entries(config).length === 0) {
    throw new Error('Configuration should not be empty!')
  }

  /**
   * Create a new validated config because the keys (patterns) might change.
   * Since the Object.reduce method already loops through each entry in the config,
   * it can be used for validating the values at the same time.
   */
  const validatedConfig = Object.entries(config).reduce((collection, [pattern, task]) => {
    /** Versions < 9 had more complex configuration options that are no longer supported. */
    if (TEST_DEPRECATED_KEYS.has(pattern)) {
      const testFn = TEST_DEPRECATED_KEYS.get(pattern)
      if (testFn(task)) {
        errors.push(
          configurationError(
            pattern,
            'Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged',
            task
          )
        )
      }

      /** Return early for deprecated keys to skip validating their (deprecated) values */
      return collection
    }

    if (
      (!Array.isArray(task) ||
        task.some((item) => typeof item !== 'string' && typeof item !== 'function')) &&
      typeof task !== 'string' &&
      typeof task !== 'function'
    ) {
      errors.push(
        configurationError(
          pattern,
          'Should be a string, a function, or an array of strings and functions',
          task
        )
      )
    }

    /**
     * A typical configuration error is using invalid brace expansion, like `*.{js}`.
     * These are automatically fixed and warned about.
     */
    const fixedPattern = withoutIncorrectBraces(pattern)
    if (fixedPattern !== pattern) {
      logger.warn(incorrectBraces(pattern, fixedPattern))
    }

    return { ...collection, [fixedPattern]: task }
  }, {})

  if (errors.length) {
    throw new Error(errors.join('\n'))
  }

  return validatedConfig
}

module.exports = validateConfig
