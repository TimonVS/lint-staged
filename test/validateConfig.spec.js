import makeConsoleMock from 'consolemock'

import validateConfig, { BRACES_REGEXP } from '../lib/validateConfig'

describe('BRACES_REGEXP', () => {
  it(`should match '*.{js}'`, () => {
    expect('*.{js}'.match(BRACES_REGEXP))
  })

  it(`should match 'file_{10}'`, () => {
    expect('file_{test}'.match(BRACES_REGEXP))
  })

  it(`should match '*.{spec\\.js}'`, () => {
    expect('*.{spec\\.js}'.match(BRACES_REGEXP))
  })

  it(`should not match '.{js,ts}'`, () => {
    expect('.{js,ts}'.match(BRACES_REGEXP))
  })

  it(`should not match 'file_{1..10}'`, () => {
    expect('file_{1..10}'.match(BRACES_REGEXP))
  })

  it(`should not match '*.\\{js\\}'`, () => {
    expect('*.\\{js\\}'.match(BRACES_REGEXP))
  })

  it(`should not match '*.\\{js}'`, () => {
    expect('*.\\{js}'.match(BRACES_REGEXP))
  })

  it(`should not match '*.{js\\}'`, () => {
    expect('*.{js\\}'.match(BRACES_REGEXP))
  })
})

describe('validateConfig', () => {
  let logger

  beforeEach(() => {
    logger = makeConsoleMock()
  })

  it('should throw and should print validation errors for invalid config 1', () => {
    const invalidConfig = 'test'

    expect(() => validateConfig(invalidConfig, logger)).toThrowErrorMatchingSnapshot()
  })

  it('should throw and should print validation errors for invalid config', () => {
    const invalidConfig = {
      foo: false,
    }

    expect(() => validateConfig(invalidConfig, logger)).toThrowErrorMatchingSnapshot()
  })

  it('should wrap function config into object', () => {
    const functionConfig = (stagedFiles) => [`eslint --fix ${stagedFiles}', 'git add`]

    expect(validateConfig(functionConfig, logger)).toEqual({
      '*': functionConfig,
    })
    expect(logger.printHistory()).toEqual('')
  })

  it('should not throw and should print nothing for valid config', () => {
    const validSimpleConfig = {
      '*.js': ['eslint --fix', 'git add'],
    }

    expect(() => validateConfig(validSimpleConfig, logger)).not.toThrow()
    expect(logger.printHistory()).toEqual('')
  })

  it('should not throw and should print nothing for function task', () => {
    const functionTask = {
      '*.js': (filenames) => {
        const files = filenames.join(' ')
        return `eslint --fix ${files} && git add ${files}`
      },
      '*.css': [(filenames) => filenames.map((filename) => `eslint --fix ${filename}`)],
    }

    expect(() => validateConfig(functionTask, logger)).not.toThrow()
    expect(logger.printHistory()).toEqual('')
  })

  it('should throw when detecting deprecated advanced configuration', () => {
    const advancedConfig = {
      chunkSize: 10,
      concurrent: false,
      globOptions: { matchBase: false },
      ignore: ['test.js'],
      linters: {
        '*.js': ['eslint'],
      },
      relative: true,
      renderer: 'silent',
      subTaskConcurrency: 10,
    }

    expect(() => validateConfig(advancedConfig, logger)).toThrowErrorMatchingSnapshot()
    expect(logger.printHistory()).toMatchSnapshot()
  })

  it('should not throw when config contains deprecated key but with valid task', () => {
    const stillValidConfig = {
      concurrent: 'my command',
    }

    expect(() => validateConfig(stillValidConfig, logger)).not.toThrow()
    expect(logger.printHistory()).toEqual('')
  })

  it('should warn about `*.{js}` and return fixed config', () => {
    const incorrectBracesConfig = {
      '*.{js}': 'eslint',
    }

    expect(() => validateConfig(incorrectBracesConfig, logger)).not.toThrow()
    expect(logger.printHistory()).toMatchSnapshot()
  })

  it('should warn about `*.{ts}{x}` and return fixed config', () => {
    const incorrectBracesConfig = {
      '*.{ts}{x}': 'eslint',
    }

    expect(() => validateConfig(incorrectBracesConfig, logger)).not.toThrow()
    expect(logger.printHistory()).toMatchSnapshot()
  })

  it('should not warn about `*.\\{js\\}`', () => {
    const incorrectBracesConfig = {
      '*.\\{js\\}': 'eslint',
    }

    expect(() => validateConfig(incorrectBracesConfig, logger)).not.toThrow()
    expect(logger.printHistory()).toEqual('')
  })
})
