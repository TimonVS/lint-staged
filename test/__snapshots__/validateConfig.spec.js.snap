// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateConfig should not throw and should print nothing for function task 1`] = `""`;

exports[`validateConfig should not throw and should print nothing for valid config 1`] = `""`;

exports[`validateConfig should not throw when config contains deprecated key but with valid task 1`] = `""`;

exports[`validateConfig should throw and should print validation errors for invalid config 1`] = `
"× Validation Error:

  Invalid value for 'foo'.

  Should be a string, a function, or an array of strings and functions.
 
  Configured value is: false
"
`;

exports[`validateConfig should throw and should print validation errors for invalid config 1 1`] = `"Configuration should be an object or a function!"`;

exports[`validateConfig should throw when detecting deprecated advanced configuration 1`] = `
"× Validation Error:

  Invalid value for 'chunkSize'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: 10

× Validation Error:

  Invalid value for 'concurrent'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: false

× Validation Error:

  Invalid value for 'globOptions'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: {matchBase: false}

× Validation Error:

  Invalid value for 'ignore'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: ['test.js']

× Validation Error:

  Invalid value for 'linters'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: {'*.js': ['eslint']}

× Validation Error:

  Invalid value for 'relative'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: true

× Validation Error:

  Invalid value for 'renderer'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: 'silent'

× Validation Error:

  Invalid value for 'subTaskConcurrency'.

  Advanced configuration has been deprecated. For more info, please visit: https://github.com/okonet/lint-staged.
 
  Configured value is: 10
"
`;

exports[`validateConfig should throw when detecting deprecated advanced configuration 2`] = `""`;

exports[`validateConfig should warn about \`*.{js}\` and return fixed config 1`] = `
"
WARN ‼ Detected incorrect braces with only single value: \`*.{js}\`. Reformatted as: \`*.js\`
"
`;

exports[`validateConfig should warn about \`*.{ts}{x}\` and return fixed config 1`] = `
"
WARN ‼ Detected incorrect braces with only single value: \`*.{ts}{x}\`. Reformatted as: \`*.tsx\`
"
`;

exports[`validateConfig should wrap function config into object 1`] = `""`;
